### PR TITLE
Implement optimistic resource quota utilization computation

### PR DESCRIPTION
**Pull request by bastian-src**
Instead of re-computing the Resource Quota utilization by iterating all Resource Quota hosts and determining their resources, an optimistic tracking of the utilized resources is implemented.

We add the table `hosts_resources` which stores the quota-related resources (CPU cores, Memory, disk space) per host. The utilized resources of a quota is calculated by iterating its assigned hosts'.host_resources respectively.

Moreover, we add the `resource_quotas_hosts` table which handles the association of hosts and quotas.

### CODE REVIEW

File: app/models/concerns/foreman_resource_quota/host_managed_extensions.rb [hunk 1]
```diff
@@ -7,16 +7,19 @@ module HostManagedExtensions
     include ForemanResourceQuota::Exceptions
 
     included do
-      validate :check_resource_quota_capacity
+      validate :verify_resource_quota_on_create, if: -> { new_record? }
+      validate :verify_resource_quota_on_change, if: -> { resource_quota_id_change_to_be_saved && !new_record? }
```
Comment (nadjaheitmann) [inline app/models/concerns/foreman_resource_quota/host_managed_extensions.rb:11]:
```suggestion
      validate :verify_resource_quota_on_host_create, if: -> { new_record? }
      validate :verify_resource_quota_on_resource_quota_change, if: -> { resource_quota_id_change_to_be_saved && !new_record? }
```

Comment (nadjaheitmann) [reply app/models/concerns/foreman_resource_quota/host_managed_extensions.rb:11]: Even better: Merge the two methods into one.

File: app/models/concerns/foreman_resource_quota/host_managed_extensions.rb [hunk 2]
```diff
@@ -113,5 +152,50 @@ def quota_assigment_optional?
...
+      end
+    end
+
+    def update_quota_with_host_resources(quota)
+      host_resources = determine_host_resources(quota.active_resources)
+      new_utilization = quota.utilization
+
+      host_resources.each do |resource_type, resource_value|
+        new_utilization[resource_type] ||= 0
+        new_utilization[resource_type] = yield(new_utilization[resource_type], resource_value, resource_type)
```
Comment (nadjaheitmann) [inline app/models/concerns/foreman_resource_quota/host_managed_extensions.rb:183]: Try to simplify the code such that it is easier to understand.

Comment (bastian-src) [reply app/models/concerns/foreman_resource_quota/host_managed_extensions.rb]: Done.

**Comment by bastian-src** [comment]
Tested locally with VMWare as compute resources and works fine. Thanks @nadjaheitmann & @Thorben-D for the review! :penguin:

**Pull request merged**